src/rgw/rgw_rest_s3.cc

```diff
@@ -5870,3 +5875,310 @@ bool rgw::auth::s3::S3AnonymousEngine::is_applicable(
+{
+  set_get_data(true);
+}
+
+RGWSelectObj_ObjStore_S3::~RGWSelectObj_ObjStore_S3()
+{
+  if (m_s3_csv_object!=0)
+    delete m_s3_csv_object;
+
+  delete s3select_syntax;
+
+  if(m_buff_header)
+    delete m_buff_header;
+}
+
+void RGWSelectObj_ObjStore_S3::encode_short(char & buff, short s , int & i)
+{
+  short x = htons(s);
+  memcpy(&buff,&x,sizeof(s));i+=sizeof(s);
+}
+
+void RGWSelectObj_ObjStore_S3::encode_int(char & buff, u_int32_t s , int & i)
+{
+  u_int32_t x = htonl(s);
+  memcpy(&buff,&x,sizeof(s));i+=sizeof(s);
+}
+
+int RGWSelectObj_ObjStore_S3::creare_header_records(char *buff)
+{
+  int i = 0;
+
+  //1
+  buff[i++] = (char)strlen(header_name_str[EVENT_TYPE]);
+  memcpy(&buff[i], header_name_str[EVENT_TYPE], strlen(header_name_str[EVENT_TYPE]));
+  i += strlen(header_name_str[EVENT_TYPE]);
+  buff[i++] = (char)7;
+  encode_short(buff[i], (short)strlen(header_value_str[RECORDS]), i);
+  memcpy(&buff[i], header_value_str[RECORDS], strlen(header_value_str[RECORDS]));
+  i += strlen(header_value_str[RECORDS]);
+
+  //2
+  buff[i++] = (char)strlen(header_name_str[CONTENT_TYPE]);
+  memcpy(&buff[i], header_name_str[CONTENT_TYPE], strlen(header_name_str[CONTENT_TYPE]));
+  i += strlen(header_name_str[CONTENT_TYPE]);
+  buff[i++] = (char)7;
+  encode_short(buff[i], (short)strlen(header_value_str[OCTET_STREAM]), i);
+  memcpy(&buff[i], header_value_str[OCTET_STREAM], strlen(header_value_str[OCTET_STREAM]));
+  i += strlen(header_value_str[OCTET_STREAM]);
+
+  //3
+  buff[i++] = (char)strlen(header_name_str[MESSAGE_TYPE]);
+  memcpy(&buff[i], header_name_str[MESSAGE_TYPE], strlen(header_name_str[MESSAGE_TYPE]));
+  i += strlen(header_name_str[MESSAGE_TYPE]);
+  buff[i++] = (char)7;
+  encode_short(buff[i], (short)strlen(header_value_str[EVENT]), i);
+  memcpy(&buff[i], header_value_str[EVENT], strlen(header_value_str[EVENT]));
+  i += strlen(header_value_str[EVENT]);
+
+  return i;
+}
+
+int RGWSelectObj_ObjStore_S3::create_message(char * buff , u_int32_t result_len,u_int32_t header_len)
+{
+  u_int32_t total_byte_len = 0;
+  u_int32_t preload_crc = 0;
+  u_int32_t message_crc = 0;
+  int i = 0;
+
+  total_byte_len = result_len + 16;
+
+  encode_int(buff[i], total_byte_len, i);
+  encode_int(buff[i], header_len, i);
+
+  crc32.reset();
+  crc32 = std::for_each( buff, buff + 8, crc32 );
+  preload_crc = crc32();
+  encode_int(buff[i], preload_crc, i);
+
+  i += result_len;
+
+  crc32.reset();
+  crc32 = std::for_each( buff, buff + i, crc32 );
+  message_crc = crc32();
+  encode_int(buff[i], message_crc, i);
+
+  return i;
+}
+
+#define PAYLOAD_LINE "\n<Payload>\n<Records>\n<Payload>\n"
+#define END_PAYLOAD_LINE "\n</Payload></Records></Payload>"
+
+int RGWSelectObj_ObjStore_S3::run_s3select(const char*query,const char*input,size_t input_length)
+{
+  int status = 0;
+  csv_object::csv_defintions csv;
+
+  m_result = "012345678901"; //12 positions for header-crc
+
+  int header_size = 0;
+  s3select_syntax->parse_query(query);
```

Comment:
we should only have to parse the query string once, right? not every time we call `run_s3select()` with a new buffer segment

Reply:
exactly,
s3-select engine wont parse again upon having AST("compiled" query).
will refactor that, to avoid confusion.

Reply:
ok thanks. i'd recommend doing all of the initialization in get_params()